### PR TITLE
Multiple problems example

### PR DESCRIPTION
**Pull request by sdatspun2**
Examples are using `causes` for consistency. Added an example for 207 and multiple problems #6 .

### CODE REVIEW

File: draft-ietf-httpapi-rfc7807bis.md [hunk 1]
```diff
@@ -130,7 +131,7 @@ Content-Language: en
  "type": "https://example.com/probs/out-of-credit",
  "title": "You do not have enough credit.",
  "detail": "Your current balance is 30, but that costs 50.",
- "instance": "/account/12345/msgs/abc",
+ "instance": "/amount/value/50",
```
Comment (asbjornu) [inline draft-ietf-httpapi-rfc7807bis.md:134]: Why the change of `instance` value? I think the value `/account/12345/msgs/abc` better captures the intent of the property than `/amount/value/50` by binding the value to a specific occurrence of an error within a specific account number.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 2]
```diff
@@ -147,22 +148,60 @@ Content-Type: application/problem+json
...
-                    {
-                      "name": "color",
-                      "reason": "must be 'green', 'red' or 'blue'"}
-                  ]
+  "type": "https://example.net/validation-error",
+  "title": "The request is invalid",
+  "causes": [
+            {
+              "detail": "must be a positive integer",
+              "instance": "/age/-50"
```
Comment (asbjornu) [inline draft-ietf-httpapi-rfc7807bis.md:156]: I don't think `/age/-50` gives a good indication of what `instance` is supposed to be used for.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 3]
```diff
@@ -147,22 +148,60 @@ Content-Type: application/problem+json
...
+            },
+            {
+              "detail": "must be 'green', 'red' or 'blue'",
+              "instance": "/profile-background-color"
```
Comment (asbjornu) [inline draft-ietf-httpapi-rfc7807bis.md:160]: I don't think `/profile-background-color` gives a good indication of what `instance` is supposed to be used for.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 4]
```diff
@@ -147,22 +148,60 @@ Content-Type: application/problem+json
...
+{
+  "type": "https://example.net/multiple-errors",
+  "title": "The request has multiple errors",
+  "causes": [
+            {
+              "type": "https://example.net/validation-error",
+              "title": "Invalid value",
+              "status": 400,
+              "detail": "must be a positive integer",
+              "instance": "/age/-50"
```
Comment (asbjornu) [inline draft-ietf-httpapi-rfc7807bis.md:183]: I don't think `/age/-50` gives a good indication of what `instance` is supposed to be used for.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 5]
```diff
@@ -147,22 +148,60 @@ Content-Type: application/problem+json
...
+            },
+            {
+              "type": "https://example.net/validation-error",
+              "title": "Invalid value",
+              "status": 400,
+              "detail": "must be 'green', 'red' or 'blue'",
+              "instance": "/profile-background-color/yellow"
```
Comment (asbjornu) [inline draft-ietf-httpapi-rfc7807bis.md:190]: I don't think `/profile-background-color/yellow` gives a good indication of what `instance` is supposed to be used for.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 6]
```diff
@@ -147,22 +148,60 @@ Content-Type: application/problem+json
...
+            },
+            {
+              "type": "https://example.net/unauthorized-error",
+              "title": "You do not have enough credit.",
+              "status": 403,
+              "detail": "Your current balance is 30, but that costs 50.",
+              "instance": "/amount/value/50",
```
Comment (asbjornu) [inline draft-ietf-httpapi-rfc7807bis.md:197]: I don't think `/amount/value/50` gives a good indication of what `instance` is supposed to be used for.

File: draft-ietf-httpapi-rfc7807bis.md (hunk 2 above)
Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md:156]: I am just trying to relate detail with instance.

File: draft-ietf-httpapi-rfc7807bis.md (hunk 1 above)
Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md:134]: Ohh, that is how you read it. I read it differently. I am trying to point out the value of "that" in "Your current balance is 30, but **that** costs 50."

Comment (mnot) [reply draft-ietf-httpapi-rfc7807bis.md:134]: +1 to a clarification, if necessary.

**Comment by dret** [comment]
On 2021-10-12 00:00, Sanjay Dalal wrote:
> - "instance": "/account/12345/msgs/abc",
> + "instance": "/amount/value/50",
> 
> Ohh, that is how you read it. I read it differently. I am trying to 
> point out the value of "that" in "Your current balance is 30, but *that* 
> costs 50."

this is a slightly confusing edit. the instance message is about the 
account not having enough funds. but it's very unlikely and probably 
impractical that this message will be availble at the URI 
"/amount/value/50".

it is more likely that the service keeps a message log somewhere along 
the lines of "/account/12345/msgs/abc" and that when accessed, this log 
message will give you all the relevant details, probably including the 
"this costs 50 but you only have 30" bit.

File: draft-ietf-httpapi-rfc7807bis.md (hunk 2 above)
Comment (dret) [reply draft-ietf-httpapi-rfc7807bis.md:156]: i think we should revert back to the made-up `msg` URI.

File: draft-ietf-httpapi-rfc7807bis.md (hunk 1 above)
Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md:134]: I agree on clarification. I thought `instance` would point to problem's occurrence in the request payload since the context here is `400`. If the status was something like 5xx, I would use `instance` to point to something that may not be in the request.

`/amount/value/50` seems easy to explain in the context of a hypothetical payment transaction vs `/account/12345/msgs/abc` where I would not know what `msgs` and `abc` mean. Is it indicating that account `12345`  has messages (`msgs`) and `abc` identifies a message instance and that message instance has a problem `Your current balance is 30, but that costs 50.`? How does this help a client of an API emitting this response?

>I think of the value of instance as the URI identifying the very application/problem+json document you're currently viewing.

Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md:134]: How about now? I have used URI fragment ids in examples and clarified the usage while describing "instance".

**Comment by dret** [comment]
On 2021-10-13 03:25, Sanjay Dalal wrote:
> How about now? I have used URI fragment ids in examples and clarified 
> the usage while describing "instance".

i don't find using instance URIs particularly helpful. imho, we should 
stick to the example of the original RFC, assuming that there is some 
persistent message storage, and showing how "instance" then could be 
used to link to that message that represents the problem instance.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 7]
```diff
@@ -147,21 +149,63 @@ Content-Type: application/problem+json
...
-                      "reason": "must be 'green', 'red' or 'blue'"}
-                  ]
-}
+ "type": "https://example.net/validation-error",
+ "title": "Your request is not valid.",
+ "status": 400,
+ "causes": [
+             {
+               "detail": "must be a positive integer",
+               "property_location": "#/age"
```
Comment (asbjornu) [inline draft-ietf-httpapi-rfc7807bis.md:158]: Although `property_location` is just added as an example extension, having it in the specification gives it weight and authority. For something like this to be added (even if only as an extension to an example) I think it needs to be thoroughly discussed in a separate issue first.

Also, I don't agree with the naming; idiomatic JSON is `lowerCamelCase` (from JavaScript which again has it from Java), not `snake_case`. The naming of the `invalid-params` property in RFC 7807 points to `kebab-case` being the condoned  naming style, but I'm not sure that's intentional. Either way, let's discuss this in a separate issue.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 8]
```diff
@@ -259,12 +259,35 @@ When "instance" contains a relative URI, it is resolved relative to the document
...
+      "amount": {
+        "value": 50,
+        "currency": "USD"
+      },
+      "age": -50,
+      "profile-background-color": "yellow"
+}
+~~~
+
+The following URI fragment identifiers provided in "instance" would evaluate to the accompanying values
```
[the remark this answers is not in the document]
Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md]: ok, i have reverted and used an extension. pl review now.

File: draft-ietf-httpapi-rfc7807bis.md (hunk 7 above)
Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md]: Opened issue #32. changed to camelCase.

**Comment by dret** [comment]
On 2021-10-16 21:45, Sanjay Dalal wrote:
> +               "property_location": "#/age"

iff we use this example we probably shouldn't use the fragment 
identifier separator.

**Comment by sdatspun2** [comment]
> On 2021-10-16 21:45, Sanjay Dalal wrote: + "property_location": "#/age"
> iff we use this example we probably shouldn't use the fragment identifier separator.

@dret Is this because application/problem+json was not perhaps registered to support fragment identifier syntax or it is just a preference? see my comment on the spec.

{{RFC6901, Section 6}} says that a given media type needs to specify JSON Pointer as its fragment identifier syntax explicitly (usually, in its registration {{RFC6838}}). Does application/problem+json support JSON Pointer's fragment identifier syntax?

**Comment by sdatspun2** [comment]
> On 2021-10-17 18:15, Sanjay Dalal wrote: @dret <https://github.com/dret> Is this because application/problem+json was not perhaps registered to support fragment identifier syntax or it is just a preference? see my comment on the spec.
> as discussed before, RFC 7807 had no intentions of pointing into a resource; "instance" has an entirely different meaning.

Got it. Since JSON pointer could be represented either as a string or a URI fragment identifier, API developer could decide. Example is using the fragment identifier, but the description of the extension does not indicate any preference.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 9]
```diff
@@ -147,21 +149,63 @@ Content-Type: application/problem+json
...
+             {
+               "detail": "must be 'green', 'red' or 'blue'",
+               "problem-pointer": "#/profile/color"
+             }
+          ]     
+  }
 ~~~
 
-Note that this requires each of the subproblems to be similar enough to use the same HTTP status code. If they do not, the 207 (Multi-Status) code {{RFC4918}} could be used to encapsulate multiple status messages.
+In this case, the validation-error problem (identified by its type URI) adds two extensions; "causes" is a JSON array that holds multiple problems and "problem-pointer" is a JSON Pointer {::comment}{{RFC6901, Section 6}} says that a given media type needs to specify JSON Pointer as its fragment identifier syntax explicitly (usually, in its registration {{RFC6838}}). Does application/problem+json support JSON Pointer's fragment identifier syntax?{:/comment} that points to the source of problem in corresponding HTTP request. Note that in this example each of the subproblems are similar enough to use the same HTTP status code. If they do not, the 207 (Multi-Status) code {{RFC4918}} could be used to encapsulate multiple status messages as shown in the example below.
```
Comment (mnot) [inline draft-ietf-httpapi-rfc7807bis.md:168]: The requirement in 6901 only applies if you're using the pointer in a URI -- you're fine here.

File: draft-ietf-httpapi-rfc7807bis.md [hunk 10]
```diff
@@ -230,6 +274,30 @@ Problem type definitions MAY extend the problem details object with additional m
...
+
+~~~
+ "#/size"             ["small", "medium", "large"]
+ "#/size/0"           "small"
+ "#/age"              -50
+ "#/profile/color"    "yellow"
+~~~
+
+Extensions like "problem-pointer" are more appropriate to use for problems associated with client side errors 4xx only.
+
```
Comment (mnot) [inline draft-ietf-httpapi-rfc7807bis.md:300]: I don't think it's a good idea to go into so much detail here, given that it's just an example. Perhaps just leave the first sentence ("Similarly...") in...

File: draft-ietf-httpapi-rfc7807bis.md (hunk 9 above)
Comment (mnot) [reply draft-ietf-httpapi-rfc7807bis.md:168]: Also, there's a tradeoff here -- if someone wants to include all possible errors in the same message, they may need to use 207, and lose any semantic transparency. Or, they could just put a subset of the errors in -- e.g., the most urgent (unauthorised, in this case).

If you want to merge this I can make a subsequent proposal to address this.

File: draft-ietf-httpapi-rfc7807bis.md (hunk 10 above)
Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md]: Removed.

File: draft-ietf-httpapi-rfc7807bis.md (hunk 9 above)
Comment (sdatspun2) [reply draft-ietf-httpapi-rfc7807bis.md]: Ok, I have removed the comment. Merging...
